### PR TITLE
feat: allow to download multiple files and folders

### PR DESCRIPTION
**Pull request by Ian2012**
### Description

This PR allows to download of multiple files in a zip file. It does so by creating a celery task with the contents to download, and it creates the zip file and uploads it to the default django storage. The task ID is returned so that the client can perform long pulling to verify the state of the task. Once the task status is `SUCCESS,` the download URL is returned as part of the task result and the file is automatically downloaded for the client in a file called `download.zip`

This implementation support folders and files.

image

### CODE REVIEW

File: filesmanager/filesmanager.py [hunk 1]
```diff
@@ -789,6 +789,39 @@ def delete_asset(self, asset_key):
         except AssetNotFoundException as e:  # pylint: disable=broad-except
             log.exception(e)
 
+    @XBlock.json_handler
+    def download_content(self, data, suffix=''):
```
Comment (mariajgrimaldi) [inline filesmanager/filesmanager.py:793]: Remember adding docstrings to methods & modules

File: filesmanager/tasks.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+from celery import shared_task
+from celery.utils.log import get_task_logger
+from django.conf import settings
+from django.core.files.storage import default_storage
+from xmodule.contentstore.django import contentstore
+
+try:
+    from opaque_keys.edx.keys import AssetKey
+except ImportError:
+    AssetKey = None
```
Comment (mariajgrimaldi) [inline filesmanager/tasks.py:13]: we can add [openedx/opaque-keys](https://github.com/openedx/opaque-keys) as a requirement

File: filesmanager/tasks.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+@shared_task(bind=True)
+def create_zip_file_task(self, contents):
+    """
+    Create temporary zip file with the contents passed as a parameter
+    and uploads it to the default storage in the downloads folder.
+
+    Returns the url of the zip file in the storage.
+    """
+    task_id = self.request.id
+    fp = tempfile.NamedTemporaryFile(delete=False, suffix=".zip")
```
Comment (mariajgrimaldi) [inline filesmanager/tasks.py:30]: can we use a different var name?

File: filesmanager/tasks.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    path = f"{DOWNLOADS_FOLDER}/{task_id}.zip"
+    with zipfile.ZipFile(fp, "w") as ziph:
+        folder_structure = get_folder_structure_from_content(contents)
+        logger.info(folder_structure)
```
Comment (mariajgrimaldi) [inline filesmanager/tasks.py:34]: are we using this?

File: filesmanager/tasks.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+        for content in folder_structure:
+            asset_key = content["asset_key"]
+            if asset_key:
```
Comment (mariajgrimaldi) [inline filesmanager/tasks.py:38]:
```suggestion
            asset_key = content["asset_key"]
            if asset_key:=content.get("asset_key"):
```

File: filesmanager/tasks.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                file_content = contentstore().find(AssetKey.from_string(asset_key))
+                ziph.writestr(zinfo_or_arcname=content["path"], data=file_content.data)
+    default_storage.save(path, fp)
+    url = default_storage.url(path)
+    logger.info("Zip file uploaded: %s", url)
+    return url
```
Comment (mariajgrimaldi) [inline filesmanager/tasks.py:44]:
```suggestion
    return default_storage.url(path)
```

File: filesmanager/tasks.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+def create_zip_file_task(self, contents):
+    """
+    Create temporary zip file with the contents passed as a parameter
+    and uploads it to the default storage in the downloads folder.
+
+    Returns the url of the zip file in the storage.
+    """
+    task_id = self.request.id
+    fp = tempfile.NamedTemporaryFile(delete=False, suffix=".zip")
+    path = f"{DOWNLOADS_FOLDER}/{task_id}.zip"
```
Comment (mariajgrimaldi) [inline filesmanager/tasks.py:31]: path of?

File: filesmanager/tasks.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        logger.info(folder_structure)
+
+        for content in folder_structure:
+            asset_key = content["asset_key"]
+            if asset_key:
+                file_content = contentstore().find(AssetKey.from_string(asset_key))
+                ziph.writestr(zinfo_or_arcname=content["path"], data=file_content.data)
+    default_storage.save(path, fp)
+    url = default_storage.url(path)
+    logger.info("Zip file uploaded: %s", url)
```
Comment (mariajgrimaldi) [inline filesmanager/tasks.py:43]: are using this log?

File: filesmanager/tasks.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    return url
+
+
+def get_folder_structure_from_content(contents, base_path="/"):
+    """
+    Generate a list of dictionaries with its path and asset_key
```
Comment (mariajgrimaldi) [inline filesmanager/tasks.py:49]:
```suggestion
    Generate a list of dictionaries each one with their path and asset_key
```

File: filesmanager/tasks.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    from a contents dictionary
+    """
+    result = []
+
+    for content in contents:
+        if content["isDir"]:
+            result.extend(
+                get_folder_structure_from_content(
+                    content["children"], base_path + content["name"] + "/"
```
Comment (mariajgrimaldi) [inline filesmanager/tasks.py:58]: can we use f strings?

File: filesmanager/tasks.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+                )
+            )
+        else:
+            result.append(
+                {
+                    "path": base_path + content["name"],
```
Comment (mariajgrimaldi) [inline filesmanager/tasks.py:64]: same comment here

File: react-app/components/FileManager/index.jsx [hunk 12]
```diff
@@ -58,6 +58,41 @@ const FileManager = (props) => {
...
+  const downloadFilesTemp = async (filesToDownload) => {
+    try {
+        const createContentData = await downloadContent({ contents: filesToDownload });
+        if (createContentData.status !== StatusCodes.OK) {
+          throw new Error('Download content has failed:  Unexpected status code');
+        }
+        let data = createContentData.data;
+        getStatus(data.task_id)
+
+        return Promise.resolve('Downloading content was successfully');
```
Comment (mariajgrimaldi) [inline react-app/components/FileManager/index.jsx:74]:
```suggestion
        return Promise.resolve('Download was successful);
```

File: react-app/components/FileManager/index.jsx [hunk 13]
```diff
@@ -58,6 +58,41 @@ const FileManager = (props) => {
...
+      } catch (error) {
+        return Promise.reject('An error has ocurred while we tried to download assets');
```
Comment (mariajgrimaldi) [inline react-app/components/FileManager/index.jsx:76]:
```suggestion
        return Promise.reject('An error has occurred while downloading assets');
```

**Review by mariajgrimaldi — CHANGES_REQUESTED** [review]

**Comment by mariajgrimaldi** [comment]
Remember adding unittests for this new implementation

**Comment by mariajgrimaldi** [comment]
@Ian2012: I tried it but it's not working for me. It stays polling the task status, do I need some setup done in my local environment?

**Comment by mariajgrimaldi** [comment]
Also, can we add a message indicating that the download is in progress? So the user waits for it

File: filesmanager/tasks.py (hunk 4 above)
Comment (Ian2012) [reply filesmanager/tasks.py:34]: The folder structure is to transform the directories tree to a list to facilitate the ZipFile creation

File: filesmanager/tasks.py (hunk 8 above)
Comment (Ian2012) [reply filesmanager/tasks.py:43]: It's for development

File: filesmanager/filesmanager.py (hunk 1 above)
Comment (Ian2012) [reply filesmanager/filesmanager.py:794]: corrected

File: react-app/components/FileManager/index.jsx [hunk 14]
```diff
@@ -58,6 +58,41 @@ const FileManager = (props) => {
     downloadFileHook(fullUrl, display_name);
   };
 
+  const downloadFiles = (filesToDownload) => {
+    downloadFilesTemp(filesToDownload)
+  };
+
+  const downloadFilesTemp = async (filesToDownload) => {
```
Comment (mariajgrimaldi) [inline react-app/components/FileManager/index.jsx:65]: why is this `Temp`?

Comment (Ian2012) [reply react-app/components/FileManager/index.jsx:65]: It's a only a wrapper, naming is hard. Any ideas?

File: .annotation_safe_list.yml [hunk 15]
```diff
@@ -0,0 +1,41 @@
+# This is a Code Annotations automatically-generated Django model safelist file.
```
Comment (mariajgrimaldi) [inline .annotation_safe_list.yml:1]: why do we need this?

File: .pii_annotations.yml [hunk 16]
```diff
@@ -0,0 +1,35 @@
+source_path: ./
```
Comment (mariajgrimaldi) [inline .pii_annotations.yml:1]: I don't think we need this file

File: filesmanager/filesmanager.py [hunk 17]
```diff
@@ -1007,28 +1038,72 @@ def delete_asset(self, asset_key):
...
+        task_result = create_zip_file_task.delay(contents)
+        return {
+            "status": "success",
+            "task_id": task_result.id
+        }
+
+    @XBlock.json_handler
+    def download_status(self, data, suffix=''):  # pylint: disable=unused-argument
+        """
+        Get the status of the async task.
```
Comment (mariajgrimaldi) [inline filesmanager/filesmanager.py:1067]: of which async task? can we be a bit more specific?

File: filesmanager/tests/test_tasks.py [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+        self.assertEqual(response.status, '200 OK')  # pylint: disable=no-member
+        body = json.loads(response.body)  # pylint: disable=no-member
+        self.assertDictEqual(body, {
+            'status': 'ERROR',
+            'message': 'Provide a list of contents to download.',
+        })
+
+    def test_xblock_download_status(self):
+        """
+        Test that the xblock returns the correct status.
```
Comment (mariajgrimaldi) [inline filesmanager/tests/test_tasks.py:131]:
```suggestion
        Test that the xblock handler used to get tasks status returns the correct information.
```

File: filesmanager/tests/test_tasks.py [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+            request = self.xblock.download_status(request)
+
+            self.assertEqual(request.status, '200 OK')  # pylint: disable=no-member
+            body = json.loads(request.body)  # pylint: disable=no-member
+            self.assertDictEqual(body, {
+                'status': 'SUCCESS',
+                'result': 'download/None.zip',
+            })
+
+    def test_xblock_download_status_not_found(self):
```
Comment (mariajgrimaldi) [inline filesmanager/tests/test_tasks.py:154]: can we add docstrings to the rest of the methods?

File: filesmanager/tests/test_tasks.py [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+        find_mock.return_value = file_content
+        mock_from_string.return_value = 'asset_key'
+
+        data = {
+            "contents": [self.xblock.directories["children"][1]],
+        }
+
+        result = create_zip_file_task(contents=data["contents"])  # pylint: disable=no-value-for-parameter
+
+        self.assertEqual(result, f"/{DOWNLOADS_FOLDER}/{None}.zip")
```
Comment (mariajgrimaldi) [inline filesmanager/tests/test_tasks.py:231]: why `None`?

File: react-app/components/FileManager/index.jsx [hunk 21]
```diff
@@ -52,13 +52,51 @@ const FileManager = (props) => {
...
     const fullUrl = port ? `${protocol}//${hostname}:${port}${url}` : `${protocol}//${hostname}${url}`;
-    downloadFileHook(fullUrl, display_name);
+    downloadFileHook(fullUrl, name, false);
   };
 
+  const downloadFiles = (filesToDownload) => {
+    downloadFilesTemp(filesToDownload)
+  };
+
+  const downloadFilesTemp = async (filesToDownload) => {
```
Comment (johnvente) [inline react-app/components/FileManager/index.jsx:67]:
```suggestion
  const downloadFilesZip = async (filesToDownload) => {
```

File: react-app/components/FileManager/index.jsx [hunk 22]
```diff
@@ -52,13 +52,51 @@ const FileManager = (props) => {
...
+        let data = createContentData.data;
+        getStatus(data.task_id)
+
+        return Promise.resolve('Download was successful');
+      } catch (error) {
+        return Promise.reject('An error has ocurred while downloading assets');
+      }
+  };
+
+  const getStatus = async (taskID) => {
```
Comment (johnvente) [inline react-app/components/FileManager/index.jsx:82]:
```suggestion
  const getStatusFromZipTask = async (taskID) => {
```

File: .annotation_safe_list.yml (hunk 15 above)
Comment (Ian2012) [reply .annotation_safe_list.yml:1]: It's for quality checks, Those fail if we don't have this file defined

File: .pii_annotations.yml (hunk 16 above)
Comment (Ian2012) [reply .pii_annotations.yml:1]: Same as above

File: filesmanager/tests/test_tasks.py (hunk 20 above)
Comment (Ian2012) [reply filesmanager/tests/test_tasks.py:231]: A celery task has a notion of self, but when calling it as a function self.request is None.

File: .annotation_safe_list.yml (hunk 15 above)
Comment (mariajgrimaldi) [reply .annotation_safe_list.yml:1]: Then we should remove those quality checks as we've done in the other repos

File: filesmanager/tests/test_tasks.py (hunk 20 above)
Comment (mariajgrimaldi) [reply filesmanager/tests/test_tasks.py:240]: I still don't get why `result` includes a None value, can't we create a mock that's closer to what happens?

File: tox.ini [hunk 23]
```diff
@@ -31,33 +30,60 @@ ignore = D101,D200,D203,D212,D215,D404,D405,D406,D407,D408,D409,D410,D411,D412,D
 match-dir = (?!migrations)
 
 [pytest]
-DJANGO_SETTINGS_MODULE = filesmanager.settings.test
-addopts = --cov filesmanager --cov-report term-missing --cov-report xml --cov-fail-under 70
+DJANGO_SETTINGS_MODULE = test_settings
+addopts = --cov filesmanager --cov tests --cov-report term-missing --cov-report xml
```
Comment (BryanttV) [inline tox.ini:34]: Why do we remove the minimum coverage?

File: filesmanager/tasks.py [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+
+@shared_task(bind=True)
+def create_zip_file_task(self, contents):
+    """
+    Create temporary zip file with the contents passed as a parameter.
+
+    Returns the url of the zip file in the storage.
+    """
+    task_id = self.request.id
+    temporary_file = tempfile.NamedTemporaryFile(delete=False, suffix=".zip")
```
Comment (mariajgrimaldi) [inline filesmanager/tasks.py:32]: what's `delete=False` for?

File: filesmanager/tests/test_filesmanager.py [hunk 25]
```diff
@@ -192,9 +192,6 @@ def test_sync_content(self):
     def test_delete_content_with_contents(self):
         """
         Check delete content JSON handler with contents.
-        Expected result:
```
Comment (mariajgrimaldi) [inline filesmanager/tests/test_filesmanager.py:195]: why are we removing this?

File: filesmanager/tests/test_tasks.py [hunk 26]
```diff
@@ -0,0 +1,243 @@
...
+            AsyncResult.return_value.status = 'SUCCESS'
+            AsyncResult.return_value.result = 'download/None.zip'
+
+            request = self.xblock.download_status(request)
+
+            self.assertEqual(request.status, '200 OK')  # pylint: disable=no-member
+            body = json.loads(request.body)  # pylint: disable=no-member
+            self.assertDictEqual(body, {
+                'status': 'SUCCESS',
+                'result': 'download/None.zip',
```
Comment (mariajgrimaldi) [inline filesmanager/tests/test_tasks.py:151]: why is this None? Can we use a better return_value for line 143?

File: filesmanager/tests/test_tasks.py [hunk 27]
```diff
@@ -0,0 +1,243 @@
...
+        Test that the xblock handler used to get tasks status returns the correct information.
+        """
+        data = {
+            "task_id": 'task_id',
+        }
+
+        request = Mock()
+        request.method = 'POST'
+        request.body = json.dumps(data).encode('utf-8')
+
```
Comment (mariajgrimaldi) [inline filesmanager/tests/test_tasks.py:140]:
```suggestion
```

File: filesmanager/tests/test_tasks.py [hunk 28]
```diff
@@ -0,0 +1,243 @@
...
+        })
+
+    def test_xblock_download_status(self):
+        """
+        Test that the xblock handler used to get tasks status returns the correct information.
+        """
+        data = {
+            "task_id": 'task_id',
+        }
+
```
Comment (mariajgrimaldi) [inline filesmanager/tests/test_tasks.py:136]:
```suggestion
```

File: filesmanager/tests/test_tasks.py [hunk 29]
```diff
@@ -0,0 +1,243 @@
...
+        self.xblock.download_content(request)
+
+        mock_create_zip_file_task.delay.assert_called_once_with(data["contents"])
+
+    def test_xblock_download_content_without_contents(self):
+        """
+        Test that the xblock returns an error when the request doesn't have the contents.
+        """
+        data = {}
+
```
Comment (mariajgrimaldi) [inline filesmanager/tests/test_tasks.py:115]:
```suggestion
```

Comment (mariajgrimaldi) [reply filesmanager/tests/test_tasks.py:115]: If we're following the https://automationpanda.com/2020/07/07/arrange-act-assert-a-pattern-for-writing-good-tests/ AAA standard, when the request setup is still part of the arrange. Try applying this to the rest of the rests

**Review by mariajgrimaldi — APPROVED** [review]
LGTM. Thanks!

File: filesmanager/tasks.py (hunk 24 above)
Comment (Ian2012) [reply filesmanager/tasks.py:32]: It doesn't automatically delete the file after closed

Comment (Ian2012) [reply filesmanager/tasks.py:32]: Updated

File: filesmanager/tests/test_filesmanager.py (hunk 25 above)
Comment (Ian2012) [reply filesmanager/tests/test_filesmanager.py:195]: Those break the docs step, can I remove all references of the docs step as we don't actually use it anywhere?

File: filesmanager/tests/test_tasks.py (hunk 20 above)
Comment (Ian2012) [reply filesmanager/tests/test_tasks.py:240]: Updated based on this answer: https://stackoverflow.com/questions/47000371/mocking-celery-self-request-attribute-for-bound-tasks-when-called-directly

**Pull request merged**